Update Chromebook blog about DDEV (#388)
ddev/ddev.com: src/content/blog/ddev-docker-chromebook.md

title: "DDEV, Docker, Chromebook!"
pubDate: 2020-05-04
summary: Lessons learned running DDEV on an inexpensive Chromebook.
modifiedDate: 2025-07-08

> modifiedComment: "Significant updates five years later in 2025... it still works! Chromebook is still a little confused about who it is, but it works as a fine Linux development environment."

Reviewer: ```suggestion
modifiedComment: "Significant updates five years later in 2025... it still works! Chromebook is still a little confused about what it is, but it works as a fine Linux development environment."
```

  - Guides
---

## Introduction

> [DDEV](http://github.com/ddev/ddev) and Docker CE work fine for local development on a Chromebook!

Reviewer 2: It reads odd to start of the article with this sentence. It would be better suited in the last paragraph of the introduction. as your verdict after the arch you've described starting with the first version of the article experimenting with the asus and the fast forward bit into 2025.

> A few years ago I wrote the first version of this article after experimenting with an 8GB [Asus C425](https://www.asus.com/us/Laptops/ASUS-Chromebook-14-C425TA/) for $329, which is a lot less any fancy Mac. It worked great, mostly because Chromebooks can run pure Debian Linux in what they call the “Terminal”.

Reviewer 3: ```suggestion
A few years ago I wrote the first version of this article after experimenting with an 8GB [Asus C425](https://www.asus.com/us/Laptops/ASUS-Chromebook-14-C425TA/) for $329, which is a lot less than any fancy Mac. It worked great, mostly because Chromebooks can run pure Debian Linux in what they call the “Terminal”.
```

Fast forward to 2025 and I saw an enticing review of the very nice 16GB Lenovo Chromebook Plus, which turns out to be an [ARM64](arm64-apple-silicon-m1-ddev-local-what-does-it-all-mean.md) machine (the same architecture as Apple Silicon). It's a much beefier machine, with great build quality and a much higher finish level, but again very nice.

Installing on a Chromebook is a reminder of how versatile DDEV is. In almost any environment that has a Docker provider you get get it going with minimal effort. That includes Windows, WSL2, Mac, Linux, Chromebook, even [Raspberry Pi](watch-ddev-local-on-arm64-raspberry-pi.md).

## Basic Chromebook Setup for DDEV

Here’s the step-by-step to set up a DDEV development environment in 2025:

1. In _Settings_ → About ChromeOS → Developers → Linux Development Environment, “Set up". (For the disk size, you'll be using Docker images, so I used about 100GB, but it's what you need that matters.)
2. In Linux, create a password for the username you’ve created by running `sudo passwd $USER`.

> 3. Do some basic updates: `sudo apt-get update && sudo apt-get upgrade -y`, then `sudo apt-get install -y git vim curl wget gnupg2 lsb-release apt-transport-https ca-certificates`.

Reviewer 3: ````suggestion
3. Do some basic updates: 

    ```shell
    sudo apt-get update && sudo apt-get upgrade -y
    sudo apt-get install -y git vim curl wget gnupg2 lsb-release apt-transport-https ca-certificates
    ```
````

Improves readbility

> 4. Install Docker CE as in the [DDEV docs](https://ddev.readthedocs.io/en/stable/users/install/docker-installation/#docker-installation-linux). You’ll be installing the [_Debian_ version](https://docs.docker.com/install/linux/docker-ce/debian/).

Reviewer 2: i've first clicked the link for the ddev docs since i thought i have to follow the instructions there. on that page was only a list of links to the different distros and the link for debian brought me to the page the second link points to for debian as well. i am not sure if dropping the first link would make sense, it is good to link the docs never the less but maybe a note that the instructions to follow are in the second link?

9. Install the Chromium browser with `sudo apt-get update && sudo apt-get install -y chromium`. You’ll also want to run `mkcert -install`. Then use the browser inside the Debian Linux container.
10. Create your project or check it out, `ddev config`, `ddev start`. Everything works, including `ddev launch`. Create a project the normal way and get to work! This is the simple path to a Drupal 11 Composer build, but there are lots of other DDEV [quickstart guides](https://ddev.readthedocs.io/en/stable/users/quickstart/).

    - `mkdir -p ~/workspace/d11 && cd ~/workspace/d11`
    - `ddev config --project-type=drupal11 --docroot=web`
    - `ddev composer create-project drupal/recommended-project`
    - `ddev launch` and go install it!
11. If you want to use PhpStorm or GoLand, they're easy enough to install in the Linux environment. You may need to install `snapd` first, (`sudo apt update && sudo apt install -y snapd`) and then install them with `sudo snap install phpstorm --classic` or `sudo snap install goland --classic`.

## Some comments about working with the Chromebook:

> - It was a pleasant experience with a nice keyboard. Outstanding performance, but less than my MacBook Air M4 It has great battery life, even with Docker running and doing lots. It has Linux (Debian), which is lovely. And the Linux side has great graphics support (GUI Linux apps work well and display with no effort). It has two USB-C ports but also a USB-A.

Reviewer 2: ```suggestion
- It was a pleasant experience with a nice keyboard. Outstanding performance, but less than my MacBook Air M4. It has great battery life, even with Docker running and doing lots. It has Linux (Debian), which is lovely. And the Linux side has great graphics support (GUI Linux apps work well and display with no effort). It has two USB-C ports but also a USB-A.
```
i "think" a period was missing after the M4, since "It" is upper case. and "...and doing lots" sounds sort of incomplete and something is missing. would it make sense to add something like "lots of tasks"?

> - Like Windows with WSL2, this is a bit schizophrenic. I had to run several tools in the Linux environment (like GoLand and PhpStorm, both worked there though as snap installs.) And I ended up running 1Password *both* in the Chrome side (browser extension) and as an app on the Linux side so I could get proper SSH Agent support from it. You have a great Linux computer that works with the nice ChromeOS desktop, and the Linux is pretty much natural, standard, no hoops to jump through.

Reviewer 2: i would avoid a term like "schizophrenic"

" I had to run several tools in the Linux environment (like GoLand and PhpStorm, both worked there though as snap installs.) And I ended up running 1Password both in the Chrome side (browser extension) and as an app on the Linux side so I could get proper SSH Agent support from it." is rather a single sentence but the "And" after the parenthesis is upper case. so it would have to be either a period or make and lower case. the latter is the advisable step i suppose. and i probably would try to rephrase the second and after the browser extension in parenthesis, maybe use something like "as well as"

> - I ran into some apps that I was unsatisfied with. Slack and [Notion.so](http://notion.so) didn’t work well as Android apps. I installed Discord as an Android app, and didn't like it much.

Reviewer 2: as the reader i wonder here what is the difference between didnt work well and didnt like much. was the didnt like much bit due to other reasons that discord not working well?

> - Of course I didn’t have the versatility of the Mac or the predictability and wide app support of Windows. But if I were doing only web development and normal editing/daily work stuff it might work out. Mainstream environments are easier. Plain Ubuntu is easier because it's just one machine, not several.

Reviewer 2: the first two sentence feel like a single one?

Author: ```suggestion
- Of course I didn’t have the versatility of the Mac or the predictability and wide app support of Windows, but if I were doing only web development and normal editing/daily work stuff it might work out. Mainstream environments are easier. Plain Ubuntu is easier because it's just one machine, not several.
```

Basically, the Chromebook-side browser doesn't innately know about the lovely Debian Linux instance that's waiting there, so it doesn't know how to connect to DDEV, which essentially appears to be on a different computer. These instructions are inspired by [How To Apply ChromeOS's Linux Container's /etc/hosts Outside of the Container in Chrome](https://chrisbeley.com/software-engineering/how-to-apply-chromeos's-linux-container's-etchosts-outside-of-the-container-in-chrome), which is a little more complex than we need for most things, but it was certainly inspiring.

These assume that you already have DDEV installed and working in the `terminal`.

1. Install the simple DNS server `dnsmasq` with `sudo apt update && sudo apt install -y dnsmasq dnsutils`.
2. Configure DDEV to allow connections from other machines with `ddev config global --router-bind-all-interfaces` and optionally `ddev config --bind-all-interfaces` for your project. (The Chromebook/Chrome side is essentially a different machine.)

> 3. The CA (Certificate Authority) that Chrome uses to authenticate certificates is normally installed with just `mkcert -install`, and you should already have done that, but remember that the Chromebook browser is completely separate from anything `mkcert` can access, so we'll import the CA manually in Chrome. In Chrome's Certificate Manager (`chrome://certificate-manager/`), go to "Local Certificates" and import from "Linux Files" the `rootCA.pem` in the directory shown by `mkcert -CAROOT`. (I didn't know how to navigate to the hidden directory `~/.local/share/mkcert` so I just copied `~/.local/share/mkcert/rootCA.pem` of of there into my `Downloads` directory.)

Reviewer 2: ```suggestion
3. The CA (Certificate Authority) that Chrome uses to authenticate certificates is normally installed with just `mkcert -install`, and you should already have done that, but remember that the Chromebook browser is completely separate from anything `mkcert` can access, so we'll import the CA manually into Chrome. In Chrome's Certificate Manager (`chrome://certificate-manager/`), go to "Local Certificates" and import from "Linux Files" the `rootCA.pem` in the directory shown by `mkcert -CAROOT`. (I didn't know how to navigate to the hidden directory `~/.local/share/mkcert` so I just copied `~/.local/share/mkcert/rootCA.pem` into my `Downloads` directory.)
```

4. Find out the IP address of the `terminal` Debian container with `dig +short penguin.lxd`. Mine was `100.115.92.204`, so I'll use that as an example.

> 5. Now we want to configure `dnsmasq` to act as a fairly normal DNS server, but also one that reports `*.ddev.site` as `100.115.92.204` (in my case). 

Reviewer 2: ```suggestion
5. Now we want to configure `dnsmasq` to act as a fairly normal DNS server, but also one that reports `*.ddev.site` as `100.115.92.204` in my case. 
```
i would drop the parenthesis. i always try to keep the number of parenthesis low.

  - Guides
---

## Introduction

> [DDEV](http://github.com/ddev/ddev) and Docker CE work fine for local development on a Chromebook!

Reviewer: ```suggestion
[DDEV](https://github.com/ddev/ddev) and Docker CE work fine for local development on a Chromebook!
```

Here’s the step-by-step to set up a DDEV development environment in 2025:

1. In _Settings_ → About ChromeOS → Developers → Linux Development Environment, “Set up". (For the disk size, you'll be using Docker images, so I used about 100GB, but it's what you need that matters.)
2. In Linux, create a password for the username you’ve created by running `sudo passwd $USER`.
3. Do some basic updates: 

    ```shell
    sudo apt-get update && sudo apt-get upgrade -y
    sudo apt-get install -y git vim curl wget gnupg2 lsb-release apt-transport-https ca-certificates
    ```
4. Install Docker CE as in the [DDEV docs](https://ddev.readthedocs.io/en/stable/users/install/docker-installation/#docker-installation-linux). You’ll be installing the [_Debian_ version](https://docs.docker.com/install/linux/docker-ce/debian/).
5. Install DDEV using the normal [Debian install instructions](https://ddev.readthedocs.io/en/stable/users/install/ddev-installation/#debianubuntu).
6. Add your user to the `docker` group by running `sudo usermod -aG docker $USER`.
7. Reboot the Debian container by running `sudo reboot`, then open “Terminal” again. You should be able to run `docker ps` successfully.

> 8. To use a browser to visit your site, you have two options. The first (preferred) is to install a browser like Chromium or Firefox and use the built-in graphics capabilities to use that browser. The second is to use the Chromebook's browser, which is far more complex. I'll add an addendum about how you can do it, but it's not easy and not recommended, but it's a fun demonstration of `dnsmasq`.

Reviewer: ```suggestion
8. To access your site in a browser, you have two options. The first (preferred) is to install a browser like Chromium or Firefox and use the system's built-in graphics to run it. The second is to use the Chromebook's browser, which is far more complex. I'll add an addendum about how you can do it, but it's not easy and not recommended, but it's a fun demonstration of `dnsmasq`.
```
